### PR TITLE
analog. add analog codec frame support

### PR DESCRIPTION
**Pull request by Pnom**

### CODE REVIEW

**Review by DarkPark — DISMISSED** [review]

File: src/analog/message.ts [hunk 1]
```diff
@@ -18,26 +18,21 @@ import getBase64FromBytes from '../utils/getBase64FromBytes.js';
...
-    /** command source binary data */
-    data: {
-        header: Uint8Array,
-        body: Uint8Array
-    },
-    /** specific command instance */
-    command: Command
+interface IMessageCommand extends ICommandBinary {
+    command: Command;
 }
```
Comment (DarkPark) [inline src/analog/message.ts:23]: how about: 

```ts
interface IMessageCommand {
    data: ICommandBinary,
    command: Command
}
```
?

File: src/analog/message.ts [hunk 2]
```diff
@@ -18,26 +18,21 @@ import getBase64FromBytes from '../utils/getBase64FromBytes.js';
...
 
-interface IMessage {
+export interface IMessage {
     commands: Array<IMessageCommand>,
     lrc: {
         expected: number | undefined,
         actual: number
     },
+    buffer: Uint8Array,
```
Comment (DarkPark) [inline src/analog/message.ts:31]:
```suggestion
    commandsData: Uint8Array,
```

?

it hints now that it shadows `commands` field

File: src/analog/commands/downlink/GetExAbsArchiveHoursMC.ts [hunk 3]
```diff
@@ -94,7 +93,12 @@ class GetExAbsArchiveHoursMC extends Command {
         buffer.setHours(hour, hours);
         buffer.setChannels(channelList.map(index => ({index} as IChannel)));
 
-        return Command.toBytes(COMMAND_ID, buffer.toUint8Array());
+        return Command.toBinary(COMMAND_ID, buffer.toUint8Array());
+    }
+
+    // returns full message - header with body
+    toBytes (): Uint8Array {
+        return this.toBinary().buffer;
```
Comment (DarkPark) [inline src/analog/commands/downlink/GetExAbsArchiveHoursMC.ts:101]: seems toBytes now does nothing but wraping toBinary

do we need it?

File: src/analog/Command.ts [hunk 4]
```diff
@@ -87,29 +93,42 @@ abstract class Command {
...
     /**
-     * Build header with body.
+     * Build header, body and complete buffer.
      *
      * @param id command id
      * @param commandData optional command binary data
      * @returns merged data
      */
-    static toBytes ( id: number, commandData?: Uint8Array ): Uint8Array {
+
```
Comment (DarkPark) [inline src/analog/Command.ts:102]:
```suggestion
```

File: src/analog/Command.ts [hunk 5]
```diff
@@ -87,29 +93,42 @@ abstract class Command {
     // }
 
     /**
-     * Build header with body.
+     * Build header, body and complete buffer.
      *
      * @param id command id
      * @param commandData optional command binary data
      * @returns merged data
```
Comment (DarkPark) [inline src/analog/Command.ts:100]: not merged

File: src/analog/message.ts [hunk 6]
```diff
@@ -144,6 +142,29 @@ export const fromBase64 = ( data: string ) => (
...
+    );
+    const actualLrc = calculateLrc(body);
+
+    return {
+        commands: [], //commandsBinary,
+        lrc: {
+            expected: actualLrc,
+            actual: actualLrc
+        },
+        buffer: mergeUint8Arrays(body, new Uint8Array([actualLrc])),
```
Comment (DarkPark) [inline src/analog/message.ts:163]: how about simpler approach?

new Uint8Array([...body, actualLrc])

here in other places

File: src/analog/message.ts [hunk 7]
```diff
@@ -144,6 +142,29 @@ export const fromBase64 = ( data: string ) => (
     fromBytes(getBytesFromBase64(data))
 );
 
+export const toMessage = ( commands: Array<Command> ): IMessage => {
+    const commandsBinary = commands.map(command => ({
+        command,
+        ...command.toBinary()
+    }));
+
+    const body = commandsBinary.reduce(
```
Comment (DarkPark) [inline src/analog/message.ts:151]: just an idea

take a look at Array.prototype.flat

Comment (Pnom) [reply src/analog/message.ts:151]: [[1], [2]].flat() will give us [1, 2]
but we have
[{buffer: [1]}, {buffer: [2]}]

File: src/analog/Command.ts (hunk 4 above)
Comment (Pnom) [reply src/analog/Command.ts]: fixed

File: src/analog/Command.ts (hunk 5 above)
Comment (Pnom) [reply src/analog/Command.ts]: comments updated

File: src/analog/message.ts (hunk 6 above)
Comment (Pnom) [reply src/analog/message.ts]: fixed

File: src/analog/message.ts (hunk 7 above)
Comment (DarkPark) [reply src/analog/message.ts:151]: `[{buffer: [1]}, {buffer: [2]}].map(item => ...item).flat()`

?

File: src/analog/message.ts (hunk 1 above)
Comment (Pnom) [reply src/analog/message.ts]: fixed

File: src/analog/message.ts (hunk 2 above)
Comment (Pnom) [reply src/analog/message.ts]: fixed

File: src/analog/commands/downlink/GetExAbsArchiveHoursMC.ts (hunk 3 above)
Comment (Pnom) [reply src/analog/commands/downlink/GetExAbsArchiveHoursMC.ts]: fixed

File: src/analog/message.ts (hunk 7 above)
Comment (Pnom) [reply src/analog/message.ts]: fixed

File: src/analog/message.ts (hunk 2 above)
Comment (Pnom) [reply src/analog/message.ts]: rename buffer to bytes

**Review by DarkPark — APPROVED** [review]

**Pull request merged**